### DISCUSSION TITLE
Bluetooth socket connection

### DISCUSSION THREAD

**Discussion by CccGm** [kenjdavidson/react-native-bluetooth-classic#263] [Q&A]
Hello,
I want to create a Bluetooth chat app using React-Native, so I have used this library but now I have an issue with how to connect Bluetooth Socket in the server and client, in that case, I have not found any socket-related connection library or this library I do not found any function. 
So, Please help me to how to I have connected that if that is not possible in react-native then inform me or give me some information.
any library, link, video, or function you know then inform me or say that is possible.

Thank you...😀

**Comment by kenjdavidson** [comment]
If you have an issue connecting to a bluetooth "server" you'll need to provide the code that you already have for both server and app, and I can maybe help work through it.

But this is such a large topic that without you providing any code, or even showing that you may have looked into this on your own to get started; there isn't much assistance that anyone can provide at this point in time.

**Reply by CccGm** [reply]
But how may i connect to server socket using this library and which of the function may I call to connect to the server or client connection,
 i haven't seen any option in a library or function to connect to the socket.

## My code is .....

```
import React, {useEffect, useState} from 'react';
import {TouchableOpacity,View,FlatList,Text} from 'react-native';
import RNBluetoothClassic, {
  BluetoothEventType,
} from 'react-native-bluetooth-classic';
import BleManager from 'react-native-ble-manager';

const Classic = () => {
  const [isEnable, setEnable] = useState(false);
  const [isScan, setScan] = useState(false);
  const [deviceList, setDeviceList] = useState(null);
  useEffect(() => {}, []);

  const List = async () => {
    await RNBluetoothClassic.list()
      .then(a => {
        console.log(a, 'list');
        setDeviceList(a);
      })
      .catch(a => console.log(a, 'eer list'));
  };

  const conn = async abc => {
    console.log(abc, 'id');
    try {
      await BleManager.connect(abc)
        .then(a => {
          console.log(a, 'true');
        })
        .catch(e => console.error(e));
    } catch (error) {
      console.error(error, 'erroe--');
    }
  };

  const StartScan = async () => {
    setScan(true);
    const devices = await RNBluetoothClassic.discoverDevices();
    if (devices != null) {
      setScan(false);
      setDeviceList(devices);
      console.log(devices, 'data scan --');
    } else {
      setScan(false);
      setDeviceList('no data found');
      console.log(devices, '--');
    }
  };

  return (
    <View style={{flex: 1}}>
      <Text
        style={{fontSize: 18, color: '#096482', marginLeft: 30, marginTop: 10}}>
        Device Bluetooth is on ? = {isEnable}
      </Text>
      {isScan == false ? (
        <TouchableOpacity
          style={{
            justifyContent: 'center',
            alignItems: 'center',
            backgroundColor: '#2699',
            padding: 15,
            margin: 25,
            marginHorizontal: 50,
            borderRadius: 15,
          }}
          onPress={StartScan}>
          <Text style={{color: 'black'}}>Click to start scan</Text>
        </TouchableOpacity>
      ) : (
        <TouchableOpacity
          style={{
            justifyContent: 'center',
            alignItems: 'center',
            backgroundColor: '#2699',
            padding: 15,
            margin: 25,
            marginHorizontal: 50,
            borderRadius: 15,
          }}
          disabled={true}>
          <Text style={{color: 'black'}}>Scaning ...</Text>
        </TouchableOpacity>
      )}
      <TouchableOpacity
        style={{
          justifyContent: 'center',
          alignItems: 'center',
          backgroundColor: '#2699',
          padding: 15,
          marginHorizontal: 50,
          borderRadius: 15,
        }}
        onPress={List}>
        <Text style={{color: 'black'}}>Click to show List</Text>
      </TouchableOpacity>
      {isScan != false || deviceList == null ? (
        <Text
          style={{justifyContent: 'center', alignSelf: 'center', fontSize: 25}}>
          Please Scan Data
        </Text>
      ) : deviceList == 'no data found' ? (
        <Text
          style={{justifyContent: 'center', alignSelf: 'center', fontSize: 25}}>
          No Data Found
        </Text>
      ) : (
        <View style={{flex: 1}}>
          <FlatList
            contentContainerStyle={{
              flexGrow: 1,
              borderWidth: 1,
              borderRadius: 10,
              margin: 10,
              padding: 10,
            }}
            data={deviceList}
            renderItem={({item}) => (
              <TouchableOpacity
                onPress={() => {
                  conn(item.id);
                }}>
                <View
                  style={{
                    borderWidth: 1,
                    margin: 3,
                    padding: 5,
                    borderRadius: 5,
                    flexDirection: 'row',
                  }}>
                  <Text style={{color: '#6c65b1', paddingHorizontal: 10}}>
                    {item.name || 'Unnamed Device'} :
                  </Text>
                  <Text style={{color: '#c54141', paddingHorizontal: 10}}>
                    {item.id}
                  </Text>
                </View>
              </TouchableOpacity>
            )}
            keyExtractor={item => item.id}
          />
        </View>
      )}
    </View>
  );
};

export default Classic;

```
Please give me the best solution for connection or socket,
Thanks for your response.. 😃😃

**Reply by kenjdavidson** [reply]
Using either the `connect` or the `connectToDevice` functions, establish an RFComm serial connection.

**Comment by CccGm** [comment]
If you have any app or example created than please share to me..

**Reply by kenjdavidson** [reply]
I have a project `bluetooth-classic-sample-apps` project that has one.

**Reply by kenjdavidson** [reply]
there are also a large number of projects that use this library https://github.com/kenjdavidson/react-native-bluetooth-classic/network/dependents

**Comment by CccGm** [comment]
> I have a project `bluetooth-classic-sample-apps` project that has one.

I have checked that but that gives me some errors,
So please also check, and thanks for giving me some of the projects.

I will check https://github.com/kenjdavidson/react-native-bluetooth-classic/network/dependents, if any queries then I will contact you please also check `` `bluetooth-classic-sample-apps ``` in this project which gives me some errors.

Thank you for your response...😉

**Comment by CccGm** [comment]
Hello,

I have created your project copy but in this server and client, both are connected.

Then after server side sends data are shown on client side, but client-side sent data is not shown on server side.

Please help with this problem...
That my project code link  https://github.com/CccGm/Bluetooth_Classic

**Reply by kenjdavidson** [reply]
You'll need to be a lot more specific, I'm not going through your full project looking for errors.  You'll need to do a little research and debugging.  If you need help I don't mind, but you need to put in a little work and give me some specific questions in order for me to even attempt helping.

**Reply by CccGm** [reply]
Apologies, I wasn't able to properly deliver it to you. I used your project file, but after establishing a connection between two devices, the messages from the server side (i.e. the device that's accepting connection) aren't displayed on the client side. It shows the message is sent but it is not received on the client side i.e. the device that is used for connection request. But the messages from the client side are being properly sent and it is received by the server as well. In summary, the messages sent from the client are displayed on the server side but the messages sent from the server side aren't displayed on the client side. Please if you have time, look into my code and address the problem.

Thank you!!


<img src="https://github.com/kenjdavidson/react-native-bluetooth-classic/assets/124134167/34dfbf9e-0993-4ba6-b8e8-6f852b7c6249" width="250" height="500">
<img src="https://github.com/kenjdavidson/react-native-bluetooth-classic/assets/124134167/33bb691c-9018-40a5-b438-0ba50e4ecfe4" width="250" height="500">

**Reply by kenjdavidson** [reply]
Looking at this quickly, it looks like it could be a bug with the sample app.

`accept` defaults the delimiter to `\r` 
`connect` defaults the delimiter to `\n` 

This is the only thing I can see that would be causing the issue with messages being delivered one way and not the other.

This is the best I can do, you'll have to debug to determine if this is actually the issue or if something else.

**Reply by CccGm** [reply]
I have changed that but didn't see any change.
so, Please tell me which problem I face at this time.

**Reply by kenjdavidson** [reply]
You're going to need to debug it yourself and provide more information.   I have neither the time nor the desire to do your debugging for your project.  If you debug it and find an issue, I'll be glad to help you work through it and fix, but you'll need to put a little work in.
